### PR TITLE
REFACTOR-3922: AST Refactor

### PR DESCRIPTION
**Pull request by marload**
## Description

#3922 

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 1]
```diff
@@ -836,33 +791,28 @@ def unbind(tensor_shares, **kwargs):
...
-
-            return results
+            return {
+                worker: [tensor_shares[worker] for tensor_shares in tensors_shares]
+                for worker in workers
+            }
 
-        @overloaded.function
-        def stack(tensors_shares, **kwargs):
+        def make_worker_dict(op, tensors_shares, **kwargs):
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:801]: can we move it (together with share_combine) at a better place out of the `def torch` overloading? and maybe rename to `apply_on_shares`   on smthg like this?

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 2]
```diff
@@ -525,41 +515,13 @@ def sub(self, shares: dict, other):
...
-                .share(
-                    *self.child.keys(),
-                    field=self.field,
-                    dtype=self.dtype,
-                    crypto_provider=self.crypto_provider,
-                    **no_wrap,
-                )
-                .child
-            )
+        other = self._convert_to_share_tensor(other)
```
[the remark this answers is not in the document]
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:518]: Could you recommend a new name for `_convert_to_share_tensor`?

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 3]
```diff
@@ -612,11 +574,12 @@ def _public_mul(self, shares, other, equation):
         assert equation == "mul" or equation == "matmul"
         cmd = getattr(torch, equation)
         if isinstance(other, dict):
-            return {
-                worker: (self.modulo(cmd(share, other[worker]))) for worker, share in shares.items()
-            }
+            f = lambda o, w: o[w]
```
[the remark this answers is not in the document]
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:577]: Thank You!

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:518]: `_convert_to_shared_tensor` is fine :)

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 1 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:801]: Not  sure as well - do yo think the multi_pointer.py could benefit from them?
If so we'll put hem in a generic place
if not, just put them in a staticmethod in the AST class but out of `def torch`

**Comment by marload** [comment]
I changed it as you suggested. Please review it 😄

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 4]
```diff
@@ -462,6 +456,28 @@ def __getitem__(self, indices):
 
     ## SECTION SPDZ
 
+    def _convert_to_shared_tensor(self, other):
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:459]: what does this method (convert to shared tensor) exactly do? and what is "other"

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 5]
```diff
@@ -474,40 +490,14 @@ def add(self, shares: dict, other):
...
 
         assert len(shares) == len(other)
 
         # matches each share which needs to be added according
         # to the location of the share
-        new_shares = {}
-        for k, v in shares.items():
-            new_shares[k] = self.modulo(other[k] + v)
-        return new_shares
+        return {k: self.modulo(other[k] + v) for k, v, in shares.items()}
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:500]: you should avoid encoding and use expressive readable names instead. for instance, k should be "worker" v should be "pointer"...

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 6]
```diff
@@ -525,41 +515,13 @@ def sub(self, shares: dict, other):
...
         assert len(shares) == len(other)
 
         # matches each share which needs to be added according
         # to the location of the share
-        new_shares = {}
-        for k, v in shares.items():
-            new_shares[k] = self.modulo(v - other[k])
-
-        return new_shares
+        return {k: self.modulo(v - other[k]) for k, v in shares.items()}
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:524]: you should avoid code repetition, because when we need to change this part of code we will need to change it in multiple places instead of one

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 7]
```diff
@@ -765,6 +719,22 @@ def mean(self, shares, **kwargs):
 
         return result
 
+    @staticmethod
+    def share_combine(tensors_shares):
```
Comment (abogaziah) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:723]: what does share_combine and apply to share do?

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 6 above)
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:524]: I agree with you. I will change it 😄

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 5 above)
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:500]: I agree with you. I will change it 😄

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 8]
```diff
@@ -462,6 +456,28 @@ def __getitem__(self, indices):
...
+        if isinstance(other, int):
+            other = torch.tensor([other], dtype=self.torch_dtype)
+
+        def share(other):
+            return other.share(
+                *self.child.keys(),
+                field=self.field,
+                dtype=self.dtype,
+                crypto_provider=self.crypto_provider,
+                **no_wrap,
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:469]: oh btw there is a new attribute which is missing (independently of your work), it's `protocol=...`!
So avoid this in the future you can use the `def get_class_attributes` method! Usage will be here: `.share(**self.get_class_attributes(), **no_wrap)`

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 4 above)
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py]: I removed this function and created a `to_shared_tensor` that only converts the tensor to shared_tensor 😃

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 6 above)
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py]: I dedupkicated these repeated codes using `apply_to_share`

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 8 above)
Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py]: I changed as you said and deduplicated all patterns used in this file using the `_to_shared_tensor`. ^^7

**Comment by marload** [comment]
@LaRiffle @abogaziah 

I applied the requested changes as much as possible. However, I am still learning the codebase of pysyft, so there may be some awkward or incorrect abstraction. If you have anything else to change, I would appreciate it if you could tell me everything!

Thank You! 😄

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 9]
```diff
@@ -387,13 +385,8 @@ def zero(self, shape=None):
...
             shape = self.shape if self.shape else [1]
-        zero = torch.zeros(*shape, dtype=self.torch_dtype).share(
-            *self.locations,
-            field=self.field,
-            dtype=self.dtype,
-            crypto_provider=self.crypto_provider,
-            **no_wrap,
-        )
+        zero = torch.zeros(*shape, dtype=self.torch_dtype)
+        zero = self._to_shared_tensor(zero, *self.locations)
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:389]: I would prefer the previous option which is more readable .share(**self.get_class_attributes(), **no_wrap) which is more readable!

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 10]
```diff
@@ -460,8 +451,24 @@ def __getitem__(self, indices):
         else:
             return self._getitem_public(indices)
 
+    def _to_shared_tensor(self, tensor, *args):
+        return tensor.share(*args, **self.get_class_attributes(), **no_wrap)
+
     ## SECTION SPDZ
 
+    def _basic_arithmetic_op(self, op, shares: dict, operand):
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:459]: Nice! Can you add a short docstring to explain? :)

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:704]: Maybe add a short docstring?

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 11]
```diff
@@ -765,6 +700,19 @@ def mean(self, shares, **kwargs):
...
+    def share_combine(tensors_shares):
+        workers = tensors_shares[0].keys()
+
+        return {
+            worker: [tensor_shares[worker] for tensor_shares in tensors_shares]
+            for worker in workers
+        }
+
+    @staticmethod
+    def apply_to_share(op, shares: dict):
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:713]: I'm not convinced we gain in readbility: i would remove this function, it's a one liner anyway

**Comment by LaRiffle** [comment]
> @LaRiffle @abogaziah
> 
> I applied the requested changes as much as possible. However, I am still learning the codebase of pysyft, so there may be some awkward or incorrect abstraction. If you have anything else to change, I would appreciate it if you could tell me everything!
> 
> Thank You! 😄

Keep going you're doing a nice job!

**Comment by marload** [comment]
@LaRiffle 
I applied the requested change. 😄

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 12]
```diff
@@ -462,6 +451,35 @@ def __getitem__(self, indices):
...
+
+        """
+        if isinstance(operand, int):
+            operand = torch.tensor([operand], dtype=self.torch_dtype)
+
+        if isinstance(operand, (torch.LongTensor, torch.IntTensor)):
+            operand = operand.share(
+                *self.child.keys(), **self.get_class_attributes(), **no_wrap
+            ).child
+        elif not isinstance(operand, dict):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:474]: Q: Couldn't this ```elif``` be moved upper? (like to be an ```elif``` for the upper ```if```)

Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:474]: Maybe we couldn't. As I changed it to what you suggested, it does not pass the test. 😄

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:474]: Do you still have the stacktrace at hand (maybe send me a message on slack)

Comment (marload) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:474]: I sent you a message!

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:474]: Ahh..I see...I read this like ```elif isinstace``` without the ```not```
